main.cc

```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
+#include <cassert>
+
+using namespace std;
+
+class String {
+public:
+    
+    String() :
+        String("", 0)
+    { }
+
+    String(const char* data) :
+        String(data, strlen(data))
+    { }
+
+    String(const String& s) :
+        String(s.data(), s.length())
+    { }
+    
+    String(const char* data, size_t length) {
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
+        m_data = new char[capacity];
+        m_capacity = capacity;
+        m_length = 1;
+    }
+    
+    String(String&& s) :
+        m_data(s.m_data),
+        m_capacity(s.capacity()),
+        m_length(s.length()) 
+    { 
+        s.clean();
+    }
+         
+    String& operator=(String s) {
+        swap(*this, s);
+        return *this;
+    }
+    
+    String& operator=(String&& s) {
+        // for x = x, ignoring more complicated cases where len/capacity differ
+        if (m_data == s.data()) {
+            return *this;
+        }
+        m_data = s.m_data;
+        m_capacity = s.capacity();
+        m_length = s.length();
+        s.clean();
+    }
+        
+    ~String() {
+        delete[] m_data;
+    }
+    
+    void reserve(unsigned capacity) {
+        if (capacity < m_capacity) { return; }
+        reallocate(capacity);
+    }
+    
+    unsigned int length() const {
+        return m_length;
+    }
+
+    unsigned int capacity() const {
+        return m_capacity;
+    }
+
+    const char* data() const {
+        return m_data;
+    }
+    
+    const char* c_str() const {
+        return m_data;
+    }
+ 
+    char& operator[] (unsigned int i) const {
+        assert(i < length());
+        return m_data[i];
+    }
+    
+    String& operator+=(const String& s) {
+        unsigned newlen = length() + s.length();
+        if (newlen + 1 >= m_capacity) {
+            String tmp(newlen);
+            tmp.add(m_data, 0);
+            tmp.add(s.m_data, length());
+            swap(*this, tmp);
+            return *this;
+        }
+        // no need for new allocation, capacity doesn't change
+        unsigned oldlen = length();
+        m_length = newlen;
+        add(s, oldlen);
+        return *this;
+    }
+    
+    char* begin() {
+        return m_data;
+    }
+     
+    char* end() {
+        return m_data + length();
+    }
+   
+private:
+    friend void swap(String& first, String& second) {
+        swap(first.m_capacity, second.m_capacity);
```

Comment:
In this particular case all the members are built-in types. But what if some of them were a user-defined type with it's own `swap`?  
Note that `swap` is ADL-based. So the common idiom looks like this:
using std::swap;
swap(left, right);

Reply:
OK. Fixed.